desc/magnetic_fields/_current_potential.py

```diff
@@ -842,3 +1042,790 @@ def nfp_loop(j, f):
+            M=max(3 * current_potential_field.M_Phi, 30),
+            N=max(3 * current_potential_field.N_Phi, 30),
+            NFP=int(eq.NFP),
+        )
+    if eval_grid is None:
+        eval_grid = LinearGrid(M=30, N=30, NFP=int(eq.NFP))
+    if normalize:
+        B_eq_surf = eq.compute("|B|", eval_grid)["|B|"]
+        # just need it for normalization, so do a simple mean
+        normalization_B = jnp.mean(B_eq_surf)
+    else:
+        normalization_B = 1
+
+    data["eval_grid"] = eval_grid
+    data["source_grid"] = source_grid
+
+    # plasma surface normal vector magnitude on eval grid
+    ne_mag = eq.compute(["|e_theta x e_zeta|"], eval_grid)["|e_theta x e_zeta|"]
+    # winding surface normal vector magnitude on source grid
+    ns_mag = current_potential_field.compute(["|e_theta x e_zeta|"], source_grid)[
+        "|e_theta x e_zeta|"
+    ]
+
+    # calculate net enclosed poloidal and toroidal currents
+    G_tot = -(eq.compute("G", grid=source_grid)["G"][0] / mu_0 * 2 * jnp.pi)
+
+    if external_field:
+        # calculate the portion of G provided by external field
+        # by integrating external toroidal field along a curve of constant theta
+        try:
+            G_ext = external_field.G
+        except AttributeError:
+            curve_grid = LinearGrid(
+                N=int(eq.NFP) * 50,
+                theta=jnp.array(jnp.pi),
+                rho=jnp.array(1.0),
+                endpoint=True,
+            )
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore", UserWarning)
+                # ignore warning from unequal NFP for grid and basis,
+                # as we don't know a-priori if the external field
+                # shares the same discrete symmetry as the equilibrium,
+                # so we will use a grid with NFP=1 to be safe
+                curve_data = eq.compute(
+                    ["R", "phi", "Z", "e_zeta"],
+                    grid=curve_grid,
+                )
+                curve_coords = jnp.vstack(
+                    (curve_data["R"], curve_data["phi"], curve_data["Z"])
+                ).T
+                ext_field_along_curve = external_field.compute_magnetic_field(
+                    curve_coords, basis="rpz", source_grid=external_field_grid
+                )
+            # calculate covariant B_zeta = B dot e_zeta from external field
+            ext_field_B_zeta = dot(ext_field_along_curve, curve_data["e_zeta"], axis=-1)
+
+            # negative sign here because with REGCOIL convention, negative G makes
+            # positive toroidal B
+            G_ext = -(
+                np.trapz(
+                    y=ext_field_B_zeta,
+                    x=curve_grid.nodes[:, 2],
+                )
+                / mu_0
+            )
+    else:
+        G_ext = 0
+
+    # G needed by surface current is the total G minus the external contribution
+    G = G_tot - G_ext
+    # TODO: what to do if p!=0 but q = 0?
```

Comment:
have not decided, it is a weird case. I think I will just set I = p*( G_tot - G_ext ) then set G=0, so that there is some toroidal current but no poloidal current. The coils will then have the necessary topology (no poloidal loops)